SignalMessaging/utils/OWSAvatarBuilder.m

```diff
@@ -50,34 +51,194 @@ + (UIImage *)buildRandomAvatarWithDiameter:(NSUInteger)diameter
+                                    drawBlock:(OWSAvatarDrawBlock)drawBlock
+{
+    OWSAssertDebug(drawBlock);
+    OWSAssertDebug(backgroundColor);
+    OWSAssertDebug(diameter > 0);
 
-    UIGraphicsBeginImageContext(srcImage.size);
+    CGRect frame = CGRectMake(0.0f, 0.0f, diameter, diameter);
 
-    CGContextRef context = UIGraphicsGetCurrentContext();
+    UIGraphicsBeginImageContextWithOptions(frame.size, NO, [UIScreen mainScreen].scale);
+    CGContextRef _Nullable context = UIGraphicsGetCurrentContext();
+    if (!context) {
+        return nil;
+    }
 
-    CGFloat width = srcImage.size.width;
+    CGContextSetFillColorWithColor(context, backgroundColor.CGColor);
+    CGContextFillRect(context, frame);
 
-    // Rotate
-    CGContextTranslateCTM(context, width / 2, width / 2);
-    CGContextRotateCTM(context, (CGFloat)M_PI_2);
-    CGContextTranslateCTM(context, -width / 2, -width / 2);
+    // Gradient
+    CGColorSpaceRef colorspace = CGColorSpaceCreateDeviceRGB();
+    CGFloat gradientLocations[] = { 0.0, 1.0 };
+    CGGradientRef _Nullable gradient = CGGradientCreateWithColors(colorspace,
+        (__bridge CFArrayRef) @[
+            (id)[UIColor colorWithWhite:0.f alpha:0.f].CGColor,
+            (id)[UIColor colorWithWhite:0.f alpha:0.15f].CGColor,
+        ],
+        gradientLocations);
+    if (!gradient) {
+        return nil;
+    }
+    CGPoint startPoint = CGPointMake(diameter * 0.5f, 0);
+    CGPoint endPoint = CGPointMake(diameter * 0.5f, diameter);
+    CGContextDrawLinearGradient(context,
+        gradient,
+        startPoint,
+        endPoint,
+        kCGGradientDrawsBeforeStartLocation | kCGGradientDrawsAfterEndLocation);
+    CFRelease(gradient);
 
-    [srcImage drawAtPoint:CGPointMake(0, 0)];
+    CGContextSaveGState(context);
+    drawBlock(context);
+    CGContextRestoreGState(context);
+
+    UIImage *_Nullable image = UIGraphicsGetImageFromCurrentImageContext();
 
-    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
     UIGraphicsEndImageContext();
+
     return image;
 }
 
-- (UIImage *)build
++ (void)drawInitialsInAvatar:(NSString *)initials
+                   textColor:(UIColor *)textColor
+                        font:(UIFont *)font
+                    diameter:(NSUInteger)diameter
+{
+    OWSAssertDebug(initials);
+    OWSAssertDebug(textColor);
+    OWSAssertDebug(font);
+    OWSAssertDebug(diameter > 0);
+
+    CGRect frame = CGRectMake(0.0f, 0.0f, diameter, diameter);
+
+    NSDictionary *textAttributes = @{
+        NSFontAttributeName : font,
+        NSForegroundColorAttributeName : textColor,
+    };
+    CGSize textSize =
+        [initials boundingRectWithSize:frame.size
+                               options:(NSStringDrawingUsesLineFragmentOrigin | NSStringDrawingUsesFontLeading)
+                            attributes:textAttributes
+                               context:nil]
+            .size;
+    // Ensure that the text fits within the avatar bounds, with a margin.
```

Comment:
This isn't in the spec, but IMO we should ensure that unexpected glyphs (e.g. in other languages) always fit neatly within the avatar bounds.